File: 240101/거울에 레이저 쏘기 2/shoot-a-laser-in-the-mirror-2.java
import java.util.Scanner;

public class Main {
    public static int x, y, dir;
    public static int[] dxs = new int[]{1, 0, -1, 0};
    public static int[] dys = new int[]{0, -1, 0, 1};

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        String[] grid = new String[n];

        for (int i = 0; i < n; i++)
            grid[i] = sc.next();
        int k = sc.nextInt();

        init(n, k);
        
        int cnt = 0;
        while (true){
            int nx = x + dxs[dir], ny = y + dys[dir];
            if (!inRange(nx, ny, n))
                break;
            cnt++;
            dir = getNextDir(nx, ny, dir, grid);
            x = nx; y = ny;
        }

        System.out.println(cnt);

    }

    private static int getNextDir(int x, int y, int dir, String[] grid) {
        if (grid[x].charAt(y) == '\\')
            return 3 - dir;
        else
            return dir ^ 1;
    }

    private static boolean inRange(int row, int col, int n) {
        return (0 <= row && row < n) && (0 <= col && col < n);
    }

    private static void init(int n, int k) {
        if (k <= n) {
            x = -1;
            y = k - 1;
            dir = 0;
        } else if (k <= 2 * n) {
            x = (k - n - 1);
            y = n;
            dir = 1;
        } else if (k <= 3 * n) {
            x = n;
            y = n - (k - 2 * n);
            dir = 2;
        } else {
            x = n - (k - 3 * n);
            y = -1;
            dir = 3;
        }
    }

}

File: 240101/거울에 레이저 쏘기 2/shoot-a-laser-in-the-mirror-2.py
n = int(input())
grid = [
    input()
    for _ in range(n)
]
k = int(input())
ans = 0

def init(k):
    if k <= n:
        return 0, k-1, 0
    elif k <= 2*n:
        return k - n - 1, n - 1, 1
    elif k <= 3*n:
        return n-1, n - (k - 2*n), 2
    else:
        return n - (k - 3*n), 0, 3
    

def move(x, y, next_dir):
    dxs, dys = [1, 0, -1, 0], [0, -1, 0, 1]
    nx, ny = x + dxs[next_dir], y + dys[next_dir]
    return nx, ny, next_dir


def in_range(x, y):
    return 0 <= x and x < n and 0 <= y and y < n


def simulate(x, y, curr_dir):
    global ans
 
    while in_range(x, y):
        if grid[x][y] == '/':
            x, y, curr_dir = move(x, y, curr_dir ^ 1)
        else:
            x, y, curr_dir = move(x, y, 3 - curr_dir)

        ans += 1
    

x, y, curr_dir = init(k)
simulate(x, y, curr_dir)
print(ans)